IO-Plugins for EMD and SEMPER formats

CodeMonkeyJan opened the pull request:
Added two io_plugins for the EMD format, a hdf5 standard proposed by Colin Ophus from Lawrence Berkeley National Lab (see http://emdatasets.lbl.gov/ not to be confused with the FEI EMD format which was developed later and is mentioned here: https://github.com/hyperspy/hyperspy/issues/698)
and a format for the image processing system Semper, which was developed by Owen Saxton. It has the ending '.unf' and is a binary file with an extensive header.

to266 commented:
Hi there, thanks for the contribution!

As always for all PRs, you will need to add tests and documentation + user guide changes. 

Furthermore, a couple of things related to hdf5 that I'm trying to work on at the moment, and if you included in your PR, that would be awesome:
 - for datasets (with actual data) `maxshape = tuple(None for _ in data.shape)` when creating them so that we can reshape it to whatever.
 - chunked storage (is actually a prerequisite of the previous point). The chunk estimation, I suppose, can be left default (i.e. `chunks=True`), as the reader should be smart enough to keep what's needed. In case you feel you can do a better job (i.e. making sure each chunk is at least one full "signal" (image / spectrum / whatever it might be)), you are welcome to give it a shot! This last request is really optional, as it is significantly more complex than just adding two copy-pasted bits of code into current dataset creation routines..

Finally, I will add a couple of general code-style comments in the code itself. It would be great if you could change it accordingly, should not take too long.

Having said all that, I haven't had a chance to actually have a look what the code does, but it's awesome that you contributed it, thank you very much!

to266 commented on hyperspy/io_plugins/emd.py:
```diff
@@ -0,0 +1,374 @@
+
+    _log = logging.getLogger(__name__)
+
+    def __init__(self, signals={}, user={}, microscope={}, sample={}, comments={}):
```
We encourage not to use new/empty dicts as default values. Instead you should write, e.g.

```python
def __init__(self, signals=None, ): # same for others if appropriate
    if signals is None:
        signals = {}
    # other default checks / code follows
```

to266 commented on hyperspy/io_plugins/emd.py:
```diff
@@ -0,0 +1,374 @@
+    comments: dictionary
+        Dictionary which contains additional commentary metadata.
+
+    '''
```
Could you change `'''` to `"""` for the documentation definitions?

to266 commented on hyperspy/io_plugins/emd.py:
```diff
@@ -0,0 +1,374 @@
+        self._log.debug('Calling _read_signal_from_group')
+        import hyperspy.api as hp
+        # Extract essential data:
+        data = group.get('data')[...]
```
You can just pass the `h5py.Dataset` as-is. It should ideally avoid loading it into memory unless we really need to. It's easy to get the numpy array later anyway

to266 commented on hyperspy/io_plugins/semper_unf.py:
```diff
@@ -0,0 +1,516 @@
+        arg_dict['IPLTYP'] = ipltyp
+        arg_dict['ICCOLN'] = iccoln
+        arg_dict['ICROWN'] = icrown
+        arg_dict['ICLAYN'] = iclayn
```
Since it's just assignment, it's more efficient to create the dictionary already full, so that it reads
```python
arg_dict = { 'data': data,
             'title': title,
             'offsets': (x0, y0, z0),
             'scales': (dx, dy, dz),
             'units': (ux, uy, uz),
             'date': date,
             'ICLASS': iclass,
             'IFORM': iform,
             'IVERSN': iversn,
             'ILABEL': ilabel,
             'IFORMAT': iformat,
             'IWP': iwp,
             'IPLTYP': ipltyp,
             'ICCOLN': iccoln,
             'ICROWN': icrown, 
             'ICLAYN': iclayn
            }
```
instead. do check for spelling / similar errors if copy-pasting

to266 commented on hyperspy/io_plugins/semper_unf.py:
```diff
@@ -0,0 +1,516 @@
+        arg_dict['IPLTYP'] = 248  # seems standard
+        arg_dict['ICCOLN'] = data.shape[2]//2 + 1
+        arg_dict['ICROWN'] = data.shape[1]//2 + 1
+        arg_dict['ICLAYN'] = data.shape[0]//2 + 1
```
same here with dict creation

to266 commented:
Finally, I think if you could split the two new formats into two separate pull-requests, it would be better as well

magnunor commented:
Very nice :)

I'm guessing the EMD-reader will only work with the newest version of EMD (0.2)? 

As mentioned in https://github.com/hyperspy/hyperspy/issues/699, it would be nice to make sure this works nicely with EMD-files made from different programs (excluding the FEI-EMD which probably needs its own reader plugin).

CodeMonkeyJan commented:
Hello everyone and sorry for the delayed reply!

Maybe I should have included a bit more information instead of just requesting a pull right out of the blue :-P, so I’ll catch up on that now:
I’m a PhD student at the Ernst Ruska-Centre for electron microscopy at the research centre in Jülich (Germany) and now have roughly three years of experience with Python. Recently, we addressed the issue of handling a vast array of different file formats and analytical tools which are present in our institute and the problems of exchanging data which stem from that. Our first idea was to rewrite our existing code in an open-source multiplatform language (obviously leading to our choice of Python), including read and write routines for the different formats. After checking out HyperSpy, we decided that we do not have to “reinvent the wheel” and instead maybe could contribute our efforts to it. The two io_plugins I wrote are a first step in this direction and would be useful for people (hopefully not only in our institute) who’d like to make the transition to Python.

First of all a big thank you for your input, I’m always eager to improve my coding style and learning from more experienced people is one of the best ways in my opinion. I mainly try to be compliant with PEP8, although I increased the max character number to 99 in my code (I hope that’s okay).
I included the `maxshape` and `chunksize` arguments during dataset creation (it now should also be passed directly, avoiding the loading into memory), removed the empty dictionaries as defaults and changed the documentation definitions from `'''` to `"""`. For the last one, is this purely an aesthetic choice or has this an actual influence on the documentation via sphinx?

I’m currently looking into splitting the pull request into two smaller requests. Up till now I only used versioning software for my PhD thesis to have a file history, handling larger structures with multiple branches and contributors is new for me. Would the best strategy be to split my fork of Hyperspy into two branches (each just adding one plugin) and then submitting two pull requests?
If yes I would do this (after adding documentation and tests for both plugins).

Thanks again :-),

Jan

to266 commented:
Hi Jan,

Yes, splitting your fork into two branches and submitting two separate pull-requests is the preferred way to do it.

One other thing that we don't strictly require but is a nice practice when using git is to rebase your branch (the one you want to merge) on hyperspy master once it is ready for merge.

That most often requires forced pushing to your github, but in the end it keeps the history "linear", which helps to make sense of it.

CodeMonkeyJan commented:
Rebasing also means that I can only request a pull for one feature at a time, right? Rebasing both io_plugins would mean that both would be in the master branch at the same time, so I would simply submit the second one after the first is accepted? Sorry if the question is stupid, as I said, I'm relatively new to git :-).
Also a small question regarding the documentation, is it enough to modify doc/api/hyperspy.io_plugins.rst to include the new plugin and let sphinx autotdoc take care of the rest or should I include something else?
Thanks!

to266 commented:
Yes, that is correct. Though you can PR both at the same time, and just rebase whichever is the second one after the first is merged - that way there would be more time for us to review the code in both branches :)

And yes, the `.rst` edit should be enough. You could try to compile the documentation yourself if you're worried that something does not work, of course.

Finally, really do not worry to PR not finished branches - it's a nice way for other people to follow the progress and advice and discuss if they spot something they are interested in, or if you have a question yourself.

CodeMonkeyJan commented:
Thanks for the help, I'm going to open a new pull request for the first plugin now :-)!
